### PR TITLE
LoadBalancer: Skip EdfScheduler creation in LoadBalancerBase if all host weights are equal.

### PR DESCRIPTION
**Pull request by antoniovicente**
Description:  Only create EdfScheduler for round-robing and least-request load balancers if hosts have different weights.  This reduces CPU usage during load balancer creation and memory usage in the common case where all hosts have the same weight.  Also, extend load_balancer_benchmark to show memory savings in cases where all weights are equal.  Partial benchmark results per RoundRobin LB instance per serving thread:
25,000 hosts
weighted %: 0 CPU time: 34.4ms memory: 1.408KB
weighted %: 50 CPU time: 53.5ms memory: 3.14728MB
weighted %: 100 CPU time: 34.5ms memory: 1.408KB
50,000 hosts
weighted %: 0 CPU time: 68.2ms memory: 1.408k
weighted %: 50 CPU time: 103ms memory: 6.29301M
weighted %: 100 CPU time: 68.5ms memory: 1.408k
In addition to memory reductions, least-request load balancers with equal weights will now use P2C for host selection instead of edf with adjusted weights.

Risk Level: Medium.  Minor changes to the behavior of least-requests load balancer.
Testing: unittest
Docs Changes: fixed existing docs
Release Notes: Added
Issue #3790, #3278

### CODE REVIEW

**Comment by mattklein123** [comment]
Drive by: this is awesome and I think this will also fix https://github.com/envoyproxy/envoy/issues/7652. @antoniovicente can you verify?

**Comment by antoniovicente** [comment]
> Drive by: this is awesome and I think this will also fix #7652. @antoniovicente can you verify?

Yes, this change removes the dependency on max_host_weight.  The dependency on host.stats().rq_active_ is still there for least-request LBs.

File: test/common/upstream/load_balancer_benchmark.cc [hunk 1]
```diff
@@ -79,10 +136,19 @@ void BM_RingHashLoadBalancerBuildRing(benchmark::State& state) {
     const uint64_t num_hosts = state.range(0);
     const uint64_t min_ring_size = state.range(1);
     RingHashTester tester(num_hosts, min_ring_size);
-    state.ResumeTiming();
```
Comment (snowp) [inline test/common/upstream/load_balancer_benchmark.cc]: why these changes to the ring hash part of this code? is this expected to change the behavior of the consistent hashing LBs as well?

File: source/common/upstream/load_balancer_impl.h [hunk 2]
```diff
@@ -346,8 +346,10 @@ class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase {
 
 protected:
   struct Scheduler {
-    // EdfScheduler for weighted LB.
-    EdfScheduler<const Host> edf_;
+    // EdfScheduler for weighted LB. The edf_ is only created when the original
+    // host weights of 2 or more hosts differ. When not present, the
+    // implementation of chooseHostOnce falls back to unweightedHostPick.
+    std::unique_ptr<EdfScheduler<const Host>> edf_;
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/upstream/load_balancer_impl.h]: I can, but I want to get your opinion about something first:
When using std::unique_ptr, memory usage reported by stats_integration_test.cc is 42838 bytes.  With absl::optional, memory usage goes up to 43126.

Relevant diffs:
diff --git a/source/common/upstream/load_balancer_impl.cc b/source/common/upstream/load_balancer_impl.cc
index 79bf87b79..a1dcd8167 100644
--- a/source/common/upstream/load_balancer_impl.cc
+++ b/source/common/upstream/load_balancer_impl.cc
@@ -651,7 +651,7 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
       return;
     }
 
-    scheduler.edf_ = std::make_unique<EdfScheduler<const Host>>();
+    scheduler.edf_ = EdfScheduler<const Host>();
 
     // Populate scheduler with host list.
     // TODO(mattklein123): We must build the EDF schedule even if all of the hosts are currently
@@ -710,7 +710,7 @@ HostConstSharedPtr EdfLoadBalancerBase::chooseHostOnce(LoadBalancerContext* cont
   // BaseDynamicClusterImpl::updateDynamicHostList, we must do a runtime pivot here to determine
   // whether to use EDF or do unweighted (fast) selection. EDF is non-null iff the original weights
   // of 2 or more hosts differ.
-  if (scheduler.edf_ != nullptr) {
+  if (scheduler.edf_.has_value()) {
     auto host = scheduler.edf_->pick();
     if (host != nullptr) {
       scheduler.edf_->add(hostWeight(*host), host);
diff --git a/source/common/upstream/load_balancer_impl.h b/source/common/upstream/load_balancer_impl.h
index 98fd0c75d..5dfd91888 100644
--- a/source/common/upstream/load_balancer_impl.h
+++ b/source/common/upstream/load_balancer_impl.h
@@ -349,7 +349,7 @@ protected:
     // EdfScheduler for weighted LB. The edf_ is only created when the original
     // host weights of 2 or more hosts differ. When not present, the
     // implementation of chooseHostOnce falls back to unweightedHostPick.
-    std::unique_ptr<EdfScheduler<const Host>> edf_;
+    absl::optional<EdfScheduler<const Host>> edf_;
   };
 
   void initialize();
diff --git a/test/integration/stats_integration_test.cc b/test/integration/stats_integration_test.cc
index ceadf07ba..e8f958373 100644
--- a/test/integration/stats_integration_test.cc
+++ b/test/integration/stats_integration_test.cc
@@ -212,7 +212,7 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithStats) {
   // 2019/07/06  7477     42742       43000   fork gauge representation to drop pending_increment_
   // 2019/07/15  7555     42806       43000   static link libstdc++ in tests
   // 2019/07/24  7503     43030       44000   add upstream filters to clusters
-  // 2019/08/13  7877     42838       44000   skip EdfScheduler creation if all host weights equal
+  // 2019/08/13  7877     43126       44000   skip EdfScheduler creation if all host weights equal
 
   // Note: when adjusting this value: EXPECT_MEMORY_EQ is active only in CI
   // 'release' builds, where we control the platform and tool-chain. So you
@@ -222,7 +222,7 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithStats) {
   // On a local clang8/libstdc++/linux flow, the memory usage was observed in
   // June 2019 to be 64 bytes higher than it is in CI/release. Your mileage may
   // vary.
-  EXPECT_MEMORY_EQ(m_per_cluster, 42838); // 104 bytes higher than a debug build.
+  EXPECT_MEMORY_EQ(m_per_cluster, 43126); // 104 bytes higher than a debug build.
   EXPECT_MEMORY_LE(m_per_cluster, 44000);
 }

File: source/common/upstream/load_balancer_impl.cc [hunk 3]
```diff
@@ -50,6 +50,20 @@ std::pair<int32_t, size_t> distributeLoad(PriorityLoad& per_priority_load,
   return {first_available_priority, total_load};
 }
 
+// Returns true if the weights of all the hosts in the HostVector are equal.
+bool originalHostWeightsAreEqual(const HostVector& hosts) {
+  if (hosts.size() <= 1) {
+    return true;
+  }
+  uint32_t weight = hosts[0]->weight();
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/upstream/load_balancer_impl.cc]: done.

File: source/common/upstream/load_balancer_impl.cc [hunk 4]
```diff
@@ -50,6 +50,20 @@ std::pair<int32_t, size_t> distributeLoad(PriorityLoad& per_priority_load,
   return {first_available_priority, total_load};
 }
 
+// Returns true if the weights of all the hosts in the HostVector are equal.
+bool originalHostWeightsAreEqual(const HostVector& hosts) {
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/upstream/load_balancer_impl.cc]: done

**Comment by antoniovicente** [comment]
> Thanks, this seems like a good change. Mostly looks good
> 
> Seems like CI is failing with
> 
> ```
> [ RUN      ] IpVersions/ClusterMemoryTestRunner.MemoryLargeClusterSizeWithStats/IPv4
> test/integration/stats_integration_test.cc:224: Failure
> Expected equality of these values:
>   m_per_cluster
>     Which is: 42838
>   43030
> ```
> 
> which implies that we're increasing the average size per cluster. Could you look into the memory increase and then we can see if we can get rid of it or if we should bump the memory test?

Took me a while to figure out why this test wasn't failing locally.  Turns out these assertions only run in CI mode.  Adjusted the test.

File: test/common/upstream/load_balancer_benchmark.cc (hunk 1 above)
Comment (antoniovicente) [reply test/common/upstream/load_balancer_benchmark.cc]: Remove changes to benchmarks for other LBs.

File: source/common/upstream/load_balancer_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (snowp) [reply source/common/upstream/load_balancer_impl.h]: I think the memory savings makes using a unique_ptr here makes more sense
